### DISCUSSION TITLE
Userscript on Button / Toggle Audio Output

### DISCUSSION THREAD

**Discussion by Thilo-90** [MiczFlor/RPi-Jukebox-RFID#1638] [Q&A] [answered]
Hi all,

i want to have a toggle function for the audio output. 
I have the sound coming from USB Soundcard Speakers and i want to be able to switch to the standard headphone jack.

I found [this Hack](https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/1389) and now i am able to switch the audio output via 
`bash RPi-Jukebox-RFID/scripts/userscripts/switch_audio_output.sh`

In the linked Hack thread it is mentioned, that the script is called via RFID, and button would be possible.

At the moment, i can only execute the script via the console command.
If i try it via Webapp /userScripts.php i get the following in my `journalctl -f`
`Nov 22 11:40:05 raspberrypi lighttpd[2499]: sudo: /home/pi/RPi-Jukebox-RFID/scripts/userscripts/switch_audio_output.sh: command not found`

I had a look in the `RPi-Jukebox-RFID/settings/gpio_settings.ini` and i can see the function calls. I could set up a button there, but what would be the function call? 
I also looked at the `components/gpio_control/function_calls.py` but, since i am new to all this coding stuff, i have no idea how to implement a new function there which would run the command `bash RPi-Jukebox-RFID/scripts/userscripts/switch_audio_output.sh`

Could someone please help me to run the script via a button press?

**Comment by manajoe** [comment]
What I recall from my memory and briefly checked in the files on my testing RPi, I edited the following files and inserted these commands:

**`RPi-Jukebox-RFID/components/gpio_control/function_calls.py`**
```def functionCallUserscript2(*args):
    function_call("{command} -c=userscript2".format(command=playout_control),
            shell=True)
```

**`RPi-Jukebox-RFID/scripts/playout_controls.sh`**
```    userscript2)
            $PATHDATA/userscripts/switch_audio_output.sh
            ;;
```

**`RPi-Jukebox-RFID/scripts/rfid_trigger_play.sh`**
```        $USERSCRIPT2)
            $PATHDATA/userscripts/switch_audio_output.sh
            ;;
```

**`RPi-Jukebox-RFID/settings/rfid_trigger_play.conf`**
```
### USERSCRIPT2
USERSCRIPT2="CARD-ID"
```

No guarantee it works, I can't test at the moment because the GPIO in my testing RPi is broken :-(

**Reply by Thilo-90** [reply]
Thanks again for your help. I'll mark my comment as answer since it has the fixes on your code and the permissions addition (in case it is needed)

**Comment by Thilo-90** [comment, marked as answer]
Good Morning,

thanks for your input @manajoe. I adjusted a bit of your code (`fuction_calls.py` has a small error)
I think it is almost working. 

**add GPIO Button**
`sudo nano RPi-Jukebox-RFID/settings/gpio_settings.ini`
```
  [JackToggle]
  enabled: True
  Type: Button
  Pin: 12
  pull_up: True
  hold_time: 0.3
  functionCall: functionCallJackToggle
```

**add function**
`nano RPi-Jukebox-RFID/components/gpio_control/function_calls.py`
```
    def functionCallJackToggle(self, *args):
        function_call("{command} -c=userscript2".format(command=self.playout_control), shell=True)
```

**add script to called function**
`nano RPi-Jukebox-RFID/scripts/playout_controls.sh`
```
    userscript2)
        # will toggle audio Jack / Speakers
        $PATHDATA/userscripts/switch_audio_output.sh
        ;;
```

in the `journalctl -f` i can see the attempt to execute the script, though `permission denied`

`Nov 23 07:58:08 raspberrypi PhonieboxGPIOControl[553]: INFO:GPIODevices.simple_button:JackToggle: execute callback
Nov 23 07:58:08 raspberrypi PhonieboxGPIOControl[553]: INFO:GPIODevices.simple_button:JackToggle: action
Nov 23 07:58:08 raspberrypi PhonieboxGPIOControl[1950]: /home/pi/RPi-Jukebox-RFID/scripts/playout_controls.sh: line 1135: /home/pi/RPi-Jukebox-RFID/scripts/userscripts/switch_audio_output.sh: Permission denied`

Any idea how to fix this? =/

EDIT: 
I got it. Read a bit about permissions. not sure which command of those triggered it, but now it works.
```
~/RPi-Jukebox-RFID/scripts/userscripts $ chmod a+r switch_audio_output.sh
~/RPi-Jukebox-RFID/scripts/userscripts $ chmod a+x switch_audio_output.sh
```
